File: src/audioWfc/constraints/NoteInKeyHardConstraint.java
package audioWfc.constraints;

import audioWfc.HigherValues;
import audioWfc.OctavedNote;
import audioWfc.Tile;
import audioWfc.constraints.grabbers.Grabber;
import audioWfc.musicTheory.Key;

public class NoteInKeyHardConstraint implements HardConstraint<OctavedNote> {
    private Grabber<Key> grabber;

    public NoteInKeyHardConstraint(Grabber<Key> grabber){
        this.grabber = grabber;
    }

    @Override
    public boolean check(Tile<OctavedNote> item, HigherValues higherValues) {
        return grabber.grab(higherValues).contains(item.getValue().getNote());
    }
}

File: vite-project/src/wfc/constraints/NoteInKeyHardConstraint.ts
import { MusicalKey } from "../../music_theory/MusicalKey"
import { OctavedNote } from "../../music_theory/Note"
import { Grabber } from "../Grabber"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { noteConstraintTypeToName } from "./constraintUtils"

export const NoteInKeyHardConstraintInit = {
	type: "NoteInKeyHardConstraint" as const,
	validByDefault: true as const,
}

export type NoteInKeyHardConstraintIR = typeof NoteInKeyHardConstraintInit

export class NoteInKeyHardConstraint {
	private grabber: Grabber<MusicalKey>
	name = noteConstraintTypeToName.get(NoteInKeyHardConstraintInit.type) as string
	constructor(grabber: Grabber<MusicalKey>) {
		this.grabber = grabber
	}

	check(tile: Tile<OctavedNote>, higherValues: HigherValues): boolean {
		const note = tile.getValue()
		const key = this.grabber(higherValues)
		return key.containsNote(note.getNote())
	}
}